design/design-sketch-oam.md

```diff
@@ -0,0 +1,334 @@
+
+The following is an example of the proposed `ApplicationConfiguration` (drawing
+from the examples included in the contemporary OAM specification) that includes
+a Crossplane `RedisCluster` resource claim in the place of an extended workload
+type, and uses a `SchedulingScope` to determine where Crossplane should deploy
+the application.
+
+```yaml
+---
+apiVersion: oam.crossplane.io/v1alpha1
+kind: SchedulingScope
+metadata:
+  namespace: default
+  name: production-europe
+  annotations:
+    description: >
+      Ensures applications are scheduled to production infrastructure in Europe.
+spec:
+  # This label selector is used to limit where this application's components
+  # may run. What exactly it selects (a Kubernetes cluster, an instance
+  # group, a serverless platform, etc) is up to the runtime.
+  selector:
+    matchLabels:
+      environment: production
+      region: europe
+---
+kind: ApplicationConfiguration
+metadata:
+  namespace: default
+  name: coolest-app
+  annotations:
+    description: A very cool cloud-native application.
+spec:
+  scopes:
+  - scopeRef:
+      apiVersion: core.oam.dev/v1alpha1
+      kind: SchedulingScope
+      name: production-europe
+  components:
+  - component:
+      apiVersion: core.oam.dev/v1alpha1
+      kind: ContainerizedWorkload
+      metadata:
+        annotations:
+          version: v1.0.0
+      spec:
+        # Under this proposal the workloadType would be limited to the core OAM
+        # types, i.e. it would define the function of the resulting workload -
+        # Server, Singleton, Worker, etc. Extended workloads are a distinct kind
+        # and thus kind: ContainerizedWorkload has no workloadSettings field.
+        workloadType: core.oam.dev/v1alpha1.Server
+        osType: linux
+        containers:
+        - name: my-twitter-bot-frontend
+          image:
+            name: example/my-twitter-bot-frontend:v1.0.0
+          resources:
+            cpu:
+              required: 1.0
+            memory:
+              required: 100MB
+          ports:
+          - name: http
+            value: 8080
+          env:
+          - name: USERNAME
+            value: cooladmin
+          - name: PASSWORD
+            valueFrom:
+              secretKeyRef:
+                name: cool-frontend
+                key: password
+          - name: BACKEND_ADDRESS
+            value: https://example.org
+          livenessProbe:
+            httpGet:
+              port: 8080
+              path: /healthz
+          readinessProbe:
+            httpGet:
+              port: 8080
+              path: /healthz
+    traits:
+    # Contemporary traits model the workload kinds to which they may apply. This
+    # could be modelled as an annotation on the CRD of each Trait kind.
+    - apiVersion: core.oam.dev/v1alpha1
+      kind: ManualScalerTrait
+      spec:
+        replicaCount: 3
+  - component:
+      apiVersion: cache.crossplane.io/v1alpha1
+      kind: RedisCluster
```

Comment:
Nits: seems this `RedisCluster` need to be changed to an extended workload spec.

Reply:
This is actually by design - one thing we were going for in this proposal is that existing Kubernetes kinds like `RedisCluster` (a Crossplane's resource claim) could effectively be "extended workloads". For a runtime like Crossplane this means we can tightly integrate our first class concepts (like resource claim) with OAM, instead of needing to define a new layer of abstraction (an extended workload kind) that would produce a resource claim behind the scenes.

Reply:
I see. 

Can you accept:

```
- component:
    instance:
      apiVersion: cache.crossplane.io/v1alpha1
      kind: RedisCluster
      spec:
        classSelector:
          matchLabels:
            environment: production
            region: europe
            ...
```
The only tiny compromise is a `instance`  field.